api/tokens.go

```diff
@@ -51,21 +81,62 @@ func (capi *census3API) getTokens(msg *api.APIdata, ctx *httprouter.HTTPContext)
 	if len(rows) == 0 {
 		return ErrNoTokens
 	}
-	// parse and encode resulting tokens
-	tokens := GetTokensResponse{Tokens: []GetTokensItem{}}
+	// init response struct with the initial pagination information and empty
+	// list of tokens
+	tokensResponse := GetTokensResponse{
+		Tokens:     []GetTokensItem{},
+		Pagination: &Pagination{PageSize: pageSize},
+	}
+	// parse current cursor
+	var currentCursor common.Address
+	if strCursor != "" {
+		currentCursor = common.HexToAddress(strCursor)
+	} else {
+		currentCursor = common.BytesToAddress(rows[0].ID)
+	}
+	// get next and previous cursors from the database and parse them to the
+	// response format
+	nextCursor, err := qtx.ListTokensNextCursor(internalCtx,
+		queries.ListTokensNextCursorParams{
+			PageCursor: currentCursor.Bytes(),
+			Limit:      pageSize,
+		})
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return ErrCantGetTokens.WithErr(err)
+	}
+	if nextCursor != nil {
+		tokensResponse.Pagination.NextCursor = common.BytesToAddress(nextCursor).String()
+	}
+	prevCursor, err := qtx.ListTokensPrevCursor(internalCtx,
+		queries.ListTokensPrevCursorParams{
+			PageCursor: currentCursor.Bytes(),
+			Limit:      pageSize,
+		})
```

Comment:
Hm, this is three queries for a single API request, which feels somewhat expensive.

Can the API client not use the first and last ID it received as cursors? For example, if the table has IDs:

```
1 2 3 4 5 6 7 8 9
```

right now you might return a response like:

```
items: [4 5 6]
next: 7
prev: 3
```

so that later you could allow the user to continue with `next` like `>= 7`.

However, you could return just `items: [4 5 6]`, and the user can give you the last ID, and you can continue the query with `> 6` instead.

basically, the pagination doesn't need to tell the user what the next and previous IDs are, outside of the query range. they can simply use "this is the smallest and largest ID I got to", and you can start the query from that point, excluding it.

Reply:
This works for the next cursor, but the problem with this approach is with the previous cursor. For example, with `pageSize=3` and `cursor=3`, the results will be `[4, 5, 6]` (all ID's greater than the `cursor`). So the smallest ID of the results, the prev cursor, will return `[5, 6, 7]` instead of `[1, 2, 3]`.

It could be solve keeping two queries, one to get the results with the current cursor, using the last result as the next cursor, and another one to get the previous cursor.

But there is another problem with knowing whether you are on the last page or not, because if you are already on the last page and it has the same number of returns as the limit of rows, the result will contain a valid next cursor (the last ID), but if you try to get the page from that cursor, no rows will be returned.